New respose body blocking: Web shells

azurit opened the pull request:
Adding new type of response body blocking - various web shells. Currently supporting only PHP web shells as i'm not able to host ASP applications (found about 2 ASP web shells). First version is able to block 24 different PHP web shells.

Huge note: This will pevent displaying of web shells only for newbie attackers as it's based on patterns found in response data (which can be changed by modifing the application). Also, gzipping of response data will bypass blocking :( was not able to workaround this. It would be really usefull if modsecurity supports something like 'decompress' transformation function.

**Finally, some help needed: I'm lost with setting of various tags in rules, can anyone help? Current tags are invalid.**

azurit commented:
Can anyone check that failed test? I don't see any error there and i also loaded that file into my web server without any problems, thanks.

azurit commented:
Should i keep setting score if i decided to change action to 'deny'?
```
    setvar:'tx.outbound_anomaly_score_pl1=+%{tx.critical_anomaly_score}',\
    setvar:'tx.anomaly_score_pl1=+%{tx.critical_anomaly_score}'"

```

fzipi commented:
You just added the `critical` value to the rules. That would be effectively blocking (denying) the request, after the correlation. Of course, unless the user changed values. Normally one critical should be enough as default for blocking.

azurit commented:
> You just added the `critical` value to the rules. That would be effectively blocking (denying) the request, after the correlation. Of course, unless the user changed values. Normally one critical should be enough as default for blocking.

But these are known 'bad things', not unknown attack detected only by some patterns. Why not do a hard-block? It's like a rule for CVE vulnerability - no reason to do any scoring, must be blocked.

fzipi commented:
Well, my idea was to be consistent with all the other `RESPONSE-*` files (I get that this might be a successful attack)

Having a second look, maybe creating a new rule for each different web shell would take way more time than just use `@pmFromFile`, like rule `id:951100`.

That would give us a more maintainable list, and will shorten the amount of rules created. What do you think?

azurit commented:
> Having a second look, maybe creating a new rule for each different web shell would take way more time than just use `@pmFromFile`, like rule `id:951100`.
> 
> That would give us a more maintainable list, and will shorten the amount of rules created. What do you think?

Would be cool except two things:
 - some of the rules are matching using `@rx`
 - `@pmFromFile` is case-**in**sensitive which i see as a problem here because there are really very low number of strings in web shells outputs with unique patters to be matched, so i find case-sensitive match important

dune73 commented:
This is waiting for Secrule_Parsing PR #6: https://github.com/coreruleset/secrules_parsing/pull/6 in order to pass the tests here.

theMiddleBlue commented on rules/RESPONSE-955-WEB-SHELLS.conf:
```diff
@@ -0,0 +1,603 @@
+    setvar:'tx.anomaly_score_pl1=+%{tx.critical_anomaly_score}'"
+
+# webadmin.php file manager
+SecRule RESPONSE_BODY "@contains <h1 style=\"margin-bottom: 0\">webadmin.php</h1>" \
```
a very impressive collection! looks good :) Just worried about this line that seems too "generic" and could lead to some FP, isn't it?

azurit replied:
Yes, that one was tricky. Maybe you will be able to find a better pattern
[webadmin.php.txt](https://github.com/coreruleset/coreruleset/files/5766009/webadmin.php.txt)

theMiddleBlue replied:
what if this line:
```
<img src="shell-webadmin.php?image=arrow"
```

theMiddleBlue replied:
> 'shell-webadmin.php' was the name of PHP file.

maybe we should shift this rule in PL2

dune73 commented:
[Meeting decision monthly chat Jan 2020](https://github.com/coreruleset/coreruleset/issues/1953#issuecomment-754192936):

 @azurit tries to find something more unique for rule 955330. If this is not possible the rule should be moved to PL2.

azurit commented:
@dune73 @lifeforms I have two more things which needs to be resolved (and i forgot about them at the meeting [there were so much staff to discuss..]):
 - Is it ok to use `deny` instead of `block`? I think we should do a hard-block here as all of these are known attacks.
 - If it's ok to use `deny`, should i still do score setting?

theMiddleBlue commented:
uhm idk if using deny instead of block, and overriding de facto the user's SecDefaultAction configuration, is a good idea (imo)

azurit commented:
@theMiddleBlue What if we have rules targeting particular CVEs? Should there be `block` too?

theMiddleBlue commented:
IMO yes. I mean if I configure my SecDefaultAction to do something on each modsecurity's block action instead of "just deny" the request (example: redirect users to a landing page, executing a lua script or add tags, etc...) I would expect to have the same behavior for all rules

dune73 commented:
You have a clear case, but so have many other rules.

If you put in "block", this will be denied by default due to the low response anomaly score threshold. If the user decides to run with a higher anomaly score threshold, then attacks going unblocked is accepted by the user.

If we step away from this policy, standard tuning techniques no longer work and false positives could have a devastating affect on new users (-> "I put the limit to 1000, but a user was blocked, when she wrote a email message in the web client about webshells. The entire email was lost. Said user hates me now. I'm removing CRS again.")

azurit commented:
I don't fully agree. There are two types of attacks:
 - one of which we are not sure is an attack, it only look like an attack (based on some patterns)
 - other of which we are 100% sure it's an attack

dune73 commented:
I do not think you can be 100% sure ever. (But this insight probably comes with age. :)

dune73 commented:
Nice use cases @theMiddleBlue!

theMiddleBlue commented:
agree with @dune73 on this, and I would like to add that many users could use the SecDefaultAction for logging purpose too (I mean for write a custom log of the modsec triggered action different than the audit or the common log).

azurit commented:
Ok i will accept it but don't agree with it. For example, in my setup, i'm using threshold of 10 only to lower the FPs (not to pass real attacks). I still want from some of the rules to do a hard block no matter on score.

azurit commented:
@theMiddleBlue For logging purpose, you shoud run in `DetectionOnly` mode.

dune73 commented:
Many people do this. But it's dangerous. You accept FPs in production. Your priority should be to tune away those FPs and lower the threshold to 5.

I never run DetectionOnly. I'm always running On, but on a new setup, I put the threshold to 10K.

theMiddleBlue commented:
yes maybe, unless you would like to log to other systems like elasticsearch or similar to get reports of modsec activities. I say it because it was my primitive setup before directly using the modsec json audit log :) but I understand you here...

azurit commented:
Should i add 'application-multi' tag? I'm missing tag descriptions.

dune73 commented:
We are not very clean with these tags and application-multi is assigned throughout the rule set. However in this particular case, I think application-multi does not apply, since it's not about the application per se. But an unwanted additional web shell application. But we are somewhat inconsistent...

azurit commented:
Moved detection of webadmin.php to PL2, changed 'deny' to 'block', fixed tags, added two new shells (malicious file managers, IDs 955340 and 955350), added notes about bypassing detection using gzipped output.

I think it's ready to merge.

dune73 commented:
Thank you for all these updates to this PR. This looks really good and it is a very welcome addition to our project.

For the record: Not providing tests is accepted for response rules (as we do not have the means / infrastructure to run them).

I am ready to merge this. But since it's a new rule file, I'd like to see more confirmations.

@fzipi / @theMiddleBlue : Ready from your perspective too?

@lifeforms : What do you think?

lifeforms commented:
I have tried out all the rules by putting the matchable contents in an HTML file and verified that they were blocked correctly.

Seeing the specificity of the patterns, I haven't passed a whole lot of traffic to it to check for FP, but I think it's good.

So merging. Thanks @azurit for a really cool addition! 🥳

dune73 commented:
-> https://twitter.com/CoreRuleSet/status/1351644573584531456

dune73 commented:
@azurit: Do you have a twitter account?

azurit commented:
> @azurit: Do you have a twitter account?

@dune73 no :smile: i am a little anti-social.

dune73 commented:
Twitter makes you anti-social, that's for sure.

Taiki-San commented:
I hadn't noticed this was ongoing, this ruleset look awesome! Thanks a ton @azurit!
I'll run it on our users shortly to get an idea of it's signal/noise ratio :)
